include/nektar_interface/projection/algorithm_types.hpp

```diff
@@ -0,0 +1,269 @@
+#pragma once
+
+#include "constants.hpp"
+#include "device_data.hpp"
+#include "shapes.hpp"
+#include <sycl/sycl.hpp>
+#include <limits>
+#include <optional>
+
+#define ROUND_UP_TO_MULTIPLE(MULTIPLE, X)                                      \
+  (MULTIPLE) * ((((X) - 1) / (MULTIPLE)) + 1)
+#define ROUND_DOWN_TO_MULTIPLE(MULTIPLE, X) (MULTIPLE) * ((X) / (MULTIPLE))
+
+namespace NESO::Project {
+namespace Private {
+template <int nmode, typename Shape>
+static inline uint64_t total_local_slots() {
+  if constexpr (Shape::dim == 2) {
+    return Shape::template local_mem_size<nmode, 0>(1) +
+           Shape::template local_mem_size<nmode, 1>(1);
+  } else if constexpr (Shape::dim == 3) {
+    return Shape::template local_mem_size<nmode, 0>(1) +
+           Shape::template local_mem_size<nmode, 1>(1) +
+           Shape::template local_mem_size<nmode, 2>(1);
+  } else {
+    static_assert(false, "Only support 2 and 3 dimensional cell types");
```

Comment:
Maybe place the static_assert before the conditional and remove this last branch.